Define a new commissioning API for Matter.framework.

bzbarsky-apple opened the pull request:
#### Summary

A new API that avoids some of the pitfalls of the existing one (for example by avoiding ambient state as much as possible, in favor of explicit state tied to a commissioning instance).

#### Testing

Unit tests to come; for now this is draft API with no implementation.

bzbarsky-apple committed:
Define a new commissioning API for Matter.framework.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCommissioningDelegate.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ * Exactly one of networks and error will be non-nil.
+ */
+- (void)comissioning:(MTRCommissioningOperation *)commissioning
```
Good catch, fixed.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.h:
```diff
@@ -0,0 +1,104 @@
+MTR_UNSTABLE_API
+@interface MTRCommissioningOperation
+/**
+ * Initialized via initWithSetupPayload.
```
Fixed.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.h:
```diff
@@ -0,0 +1,104 @@
+ *
+ * The deviceAttestationDelegate property of MTRCommissioningParameters will be
+ * ignored. Device attestation notifications will be delivered to any
+ * registereded MTRCommissioningDelegates instead.
```
Done.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.h:
```diff
@@ -0,0 +1,104 @@
+
+/**
+ * Whether to perform a network scan as part of commissioning, if network
+ * commisisoning is required.  Defaults to NO.
```
Fixed.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningDelegate.h:
```diff
@@ -0,0 +1,112 @@
+ * for commissionees that are already on-network this notification will not
+ * happen.
+ */
+- (void)commissioningProvisionedNetworkCredentials:(MTRCommissioningOperation *)commissioning;
```
We could have some generic "progress" callback that just provides the MTRCommissioningOperation and a step enum?

bzbarsky-apple replied:
Yeah.... I need to think about what should go in that enum.  But it would probably be a good idea given some other use cases Home has for knowing progress here.

bzbarsky-apple replied:
Done.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.h:
```diff
@@ -0,0 +1,65 @@
+ * Stop commissioning.  This will typically result in
+ * commissioning:failedWithError: callbacks to delegates.
+ */
+- (void)stop;
```
Should this return a BOOL to indicate if we stopped? So if e.g. the call was made too late and we were already done the result will be NO.

bzbarsky-apple replied:
Done.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRDeviceController.h:
```diff
@@ -162,6 +162,10 @@ MTR_AVAILABLE(ios(16.1), macos(13.0), watchos(9.1), tvos(16.1))
  * Commission the node with the given node ID.  The node ID must match the node
  * ID that was used to set up the commissioning session.
+ *
+ * The shouldScanWiFiNetworks and shouldScanThreadNetworks members of
```
I can't find these in the current MTRCommissioningParameters?

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.h:
```diff
@@ -0,0 +1,65 @@
+ * The provided delegate will be notified about various things as commissioning
+ * proceeds.  The calls into the delegate will happen on the provided queue.
+ */
+- (nullable instancetype)initWithParameters:(MTRCommissioningParameters *)parameters
```
Did you consider making this a subclass of MTRCommissioningParameters so the individual properties can be directly set on the operation. (Or factoring out a base-class that only contains the ones that are actually applicable to this commissioning API.)

If we're sticking with the API as is, we should clarify if the parameters can be modified after the operation is created, and/or after it's started

bzbarsky-apple replied:
I did consider that, but ran into some issues (e.g. we want parameters copyable, but this class is obviously not copyable).

And setting things directly on the operation is a bit of a problem, because in fact you can't change them after you start, and having an API like that seemed weird.

I'll add documentation saying the parameters cannot be modified after init, which is the current setup.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningDelegate.h:
```diff
@@ -0,0 +1,112 @@
+ *
+ * Exactly one of networks and error will be non-nil.
+ */
+- (void)commissioning:(MTRCommissioningOperation *)commissioning
```
We could expose the network interface info in MTRCommissioneeInfo, pretty sure the underlying C++ object already has that to a degree.

bzbarsky-apple replied:
Fundamentally right now: the C++ SDK does not support that, really, unless we allowed providing them during the device attestation callback or something.

That is, the only places we currently have pauses in the C++ commissioning state machine are device attestation and network scans. 

Longer-term, I am hoping to add a "we have no credentials, but were not asked to scan, just ask for credentials now" thing to the C++ SDK, but I had to cut off the scope for this change somewhat to make it sane.

bzbarsky-apple replied:
Tracking this in https://github.com/project-chip/connectedhomeip/pull/40778 but for now the information is exposed (in a somewhat ugly way) in "attributes", so not blocking on this part.

bzbarsky-apple committed:
Address review comments.

bzbarsky-apple committed:
Improve some logging and API documentation.

bzbarsky-apple committed:
Mark new APIs as provisionally available.

bzbarsky-apple committed:
Ensure we trigger commissioningComplete in all cases when commissining stops.

When cancelCommissioningForNodeID is called after PASE establishment but before
starting the commissioning state machine, we were not triggering
commissioningComplete callbacks on the delegate.  We should be doing that.

bzbarsky-apple committed:
Address review comment: more general progress notifications.

bzbarsky-apple committed:
Address review comments.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceController_Concrete.mm:
```diff
@@ -952,6 +1042,51 @@ - (BOOL)setupCommissioningSessionWithDiscoveredDevice:(MTRCommissionableBrowserR
+{
+    auto * currentCommissioning = self.currentCommissioning;
+    if (commissioning != currentCommissioning) {
+        MTR_LOG_ERROR("%@ trying to start comissioning %@ but current commmissioning is %@", self, commissioning, currentCommissioning);
```
Fixed.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningDelegate.h:
```diff
@@ -0,0 +1,123 @@
+#import <Matter/MTRMetrics.h>
+#import <Matter/MTRStructsObjc.h>
+
+@class MTRCommissioningOperation; // Can't be imported, since it needs to reference us.
```
Unless you're sub-classing or there are typedefs or similar that you need, I would consider forward declaring classes instead of importing the header as the preferred style anyway.

bzbarsky-apple replied:
OK.  Comment removed.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.h:
```diff
@@ -0,0 +1,74 @@
++ (instancetype)new NS_UNAVAILABLE;
+
+/**
+ * Prepare to do a new commissioning of a device with the given parameters and
```
```suggestion
 * Prepare to commission a device with the given parameters and
```

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.mm:
```diff
@@ -0,0 +1,490 @@
+using namespace chip::Tracing::DarwinFramework;
+
+@interface MTRCommissioningOperationDeviceAttestationDelegate : NSObject <MTRDeviceAttestationDelegate>
+@property (nonatomic, readwrite, weak) MTRCommissioningOperation * commissioningOperation;
```
```suggestion
@property (nonatomic, weak) MTRCommissioningOperation * commissioningOperation;
```

bzbarsky-apple replied:
Done.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.mm:
```diff
@@ -0,0 +1,490 @@
+    _delegate = delegate;
+    _delegateQueue = queue;
+
+    // Set an attestation delegate on the parameters, but avoid a reference
```
This could be moved into startWithController?

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningDelegate_Internal.h:
```diff
@@ -0,0 +1,65 @@
+         statusUpdate:(MTRCommissioningStatus)status;
+
+/**
+ * Callback that allows piping the nodeID through to
```
Can we just make the node ID available as an internal property on MTRCommissioningOperation instead of having these two special callbacks?

bzbarsky-apple replied:
This is quite difficult because of the mutating node ID (across AddNOC) business.  It might be doable if we mutate that internal property as needed, but it's more fragile and less obviously correct than just providing the information we already have.  

As a followup we can look into doing that and getting rid of these internal bits, but for now I wanted to minimize risk.

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.mm:
```diff
@@ -0,0 +1,490 @@
+        // our state on the MTRDeviceController.  But we want this part to come
+        // after the delegate notification when the delegate is the controller
+        // itself.  Try to figure out a better way to do this.
+        [strongController commissioningDone:self];
```
If you place this call after the dispatch_async, and the controller itself dispatches to its own queue, wouldn't this still ensure this work happens after the delegate callbacks are performed (i.e. two dispatch_async to the same queue)

ksperling-apple commented on src/darwin/Framework/CHIP/MTRCommissioningOperation.mm:
```diff
@@ -0,0 +1,490 @@
+    MTR_LOG("%@ Device commissioning failed with controller %@ metrics %@", self, strongController, metrics);
+
+    id<MTRCommissioningDelegate> strongDelegate = _delegate;
+    id<MTRCommissioningDelegate_Internal> strongInternalDelegate = [self _internalDelegate];
```
Why are you distinguishing `strongDelegate` and `strongInternalDelegate` here? `_internalDelegate` just performs an unchecked downcast, so they are always going to be the same pointer.

woody-apple commented on src/darwin/Framework/CHIP/MTRCommissioneeInfo.h:
```diff
@@ -55,7 +55,7 @@ MTR_AVAILABLE(ios(18.4), macos(15.4), watchos(11.4), tvos(18.4))
  * 2) The FeatureMap attributes of all Network Commissioning clusters on the commissionee.
  */
-@property (nonatomic, copy, readonly, nullable) NSDictionary<MTRAttributePath *, NSDictionary<NSString *, id> *> * attributes MTR_UNSTABLE_API;
+@property (nonatomic, copy, readonly, nullable) NSDictionary<MTRAttributePath *, NSDictionary<NSString *, id> *> * attributes MTR_PROVISIONALLY_AVAILABLE;
```
No, we're using MTR_PROVISIONALLY_AVAILABLE for now until we have something that's better.

woody-apple commented:
Leaving this for @bzbarsky-apple tomorrow, but we need to merge this tomorrow...

bzbarsky-apple committed:
Address review comments.

bzbarsky-apple committed:
Fix starting commissioning from commissioning end delegate notifications to work reliably.
